### PR TITLE
New action Blacklist URL in Zscaler plugin

### PR DESCRIPTION
**Pull request by r7-kszczepanskagorna**
## Proposed Changes
New action Blacklist URL in Zscaler plugin

### Description

Describe the proposed changes:

  -

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [ ] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [ ] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [ ] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.0. Step name: blacklist_url\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.0 --debug run < tests/blacklist_url_add.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: Invalid urls [tesbadtestbad]",
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.0. Step name: blacklist_url\nAn error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: Invalid urls [tesbadtestbad]\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.3-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 307, in handle_step\n    output = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.3-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 428, in start_step\n    output = func(params)\n  File \"/usr/local/lib/python3.8/site-packages/zscaler_rapid7_plugin-1.1.0-py3.8.egg/icon_zscaler/actions/blacklist_url/action.py\", line 23, in run\n    Output.SUCCESS: self.connection.client.blacklist_url(blacklist_step, params.get(Input.URLS))\n  File \"/usr/local/lib/python3.8/site-packages/zscaler_rapid7_plugin-1.1.0-py3.8.egg/icon_zscaler/util/api.py\", line 25, in blacklist_url\n    response = self.authenticated_call(\n  File \"/usr/local/lib/python3.8/site-packages/zscaler_rapid7_plugin-1.1.0-py3.8.egg/icon_zscaler/util/api.py\", line 63, in authenticated_call\n    return self._call_api(\n  File \"/usr/local/lib/python3.8/site-packages/zscaler_rapid7_plugin-1.1.0-py3.8.egg/icon_zscaler/util/api.py\", line 98, in _call_api\n    raise PluginException(preset=PluginException.Preset.UNKNOWN, data=response.json().get(\"message\", response.text))\ninsightconnect_plugin_runtime.exceptions.PluginException: An error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: Invalid urls [tesbadtestbad]\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.0 --debug run < tests/blacklist_url_add_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.0. Step name: blacklist_url\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.0 --debug run < tests/blacklist_url_list_add.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.0. Step name: blacklist_url\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.0 --debug run < tests/blacklist_url_remove.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.0. Step name: blacklist_url\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.0 --debug run < tests/blacklist_url_remove_list.json
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 7558.085ms

```

<summary>
icon-validate --all .
</summary>
</details>

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../tools/Makefiles/Helpers.mk ../tools/Makefiles/Colors.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 865.598ms

[*] Validating spec with js-yaml
[SUCCESS] Passes js-yaml spec check


[*] Validating markdown...
[SUCCESS] Passes markdown linting

[*] Validating python files for style...
[SUCCESS] Passes flake8 linting

[*] Validating python files for security vulnerabilities...
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.7.3
Run started:2020-09-17 18:50:31.142847

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 412
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
[SUCCESS] Passes bandit security checks

[*] Checking for typos with Misspell...
[SUCCESS] Passes Misspell check

```

<summary>
make validate
</summary>
</details>

### CODE REVIEW

File: zscaler/help.md [hunk 1]
```diff
@@ -5,19 +5,20 @@
 # Key Features
 
 * Lookup categories for a given URL
+* Add or remove an URLs from blacklist
```
Comment (jschipp-r7) [inline zscaler/help.md:8]:
```suggestion
* Add or remove URLs from blacklist
```

File: zscaler/plugin.spec.yaml [hunk 2]
```diff
@@ -75,3 +75,26 @@ actions:
         description: Information about given URLs
         type: "[]url_categorization"
         required: true
+  blacklist_url:
+    title: Blacklist URL
+    description: Add or remove an URLs from blacklist
```
Comment (jschipp-r7) [inline zscaler/plugin.spec.yaml:80]:
```suggestion
    description: Add or remove URLs from blacklist. This blocks a URL from being accessed
```

File: zscaler/plugin.spec.yaml [hunk 3]
```diff
@@ -75,3 +75,26 @@ actions:
...
+    input:
+      urls:
+        title: URLs
+        description: The given set of URLs or one URL to add or remove from blacklist
```
Comment (jschipp-r7) [inline zscaler/plugin.spec.yaml:84]:
```suggestion
        description: A given set of one or more URLs to update in the blacklist
```

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]
Can we improve error handling to not print the object but extract the message so it's printed nicely for users?

 Response was: {\"code\":\"INVALID_INPUT_ARGUMENT\",\"message\":\"Invalid urls [tesbadtestbad]\"}",

File: zscaler/help.md [hunk 4]
```diff
@@ -5,19 +5,20 @@
...
 
-
 # Documentation
+
 ## Setup
 
 To locate your base URI and key:
 
-1. Log in to the ZIA Admin Portal using your admin credentials. 
+1. Log in to the ZIA Admin Portal using your admin credentials.
```
Comment (mrinehart-r7) [inline zscaler/help.md:21]: Is this the same URL for any customer? Ie. the URL doesn't change depending on the organization or anything? If so, can you link that URL here?

File: zscaler/help.md [hunk 5]
```diff
@@ -44,11 +45,50 @@ Example input:
...
+
 ```
+
 ## Technical Details
 
 ### Actions
 
+#### Blacklist URL
+
+This action is used to add or remove an URLs from blacklist.
```
Comment (mrinehart-r7) [inline zscaler/help.md:57]:
```suggestion
This action is used to add or remove  URLs from a blacklist.
```

File: zscaler/plugin.spec.yaml [hunk 6]
```diff
@@ -75,3 +75,26 @@ actions:
         description: Information about given URLs
         type: "[]url_categorization"
         required: true
+  blacklist_url:
+    title: Blacklist URL
+    description: Add or remove URLs from blacklist. This blocks a URL from being accessed
```
Comment (mrinehart-r7) [inline zscaler/plugin.spec.yaml:80]:
```suggestion
    description: Blacklist or unblacklist a URL
```

File: zscaler/plugin.spec.yaml [hunk 7]
```diff
@@ -75,3 +75,26 @@ actions:
...
+    input:
+      urls:
+        title: URLs
+        description: A given set of one or more URLs to update in the blacklist
+        type: "[]string"
+        required: true
+        example: ["example.com", "rapid7.com"]
+      blacklist_state:
+        title: Blacklist State
+        description: Set true to add an URL to blacklist, set false to remove URL from blacklist
```
Comment (mrinehart-r7) [inline zscaler/plugin.spec.yaml:90]:
```suggestion
        description: True to blacklist a URL, false to unblacklist a URL
```

File: zscaler/plugin.spec.yaml [hunk 8]
```diff
@@ -75,3 +75,26 @@ actions:
...
+        type: boolean
+        required: false
+        default: true
+        example: true
+    output:
+      success:
+        title: Success
+        description: Information about add or remove URLs from blacklist
```
Comment (mrinehart-r7) [inline zscaler/plugin.spec.yaml:98]:
```suggestion
        description: Whether or not the request succeeded
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

File: zscaler/help.md (hunk 4 above)
Comment (r7-kszczepanskagorna) [reply zscaler/help.md:21]: For now is 5 different URLs. Unfortunately I'm not found a way to make it easier for a customer (only an enum, but if they add some URL it will make some problems).

Comment (r7-kszczepanskagorna) [reply zscaler/help.md:21]: https://help.zscaler.com/zia/api-getting-started
In section "Retrieve your base URI and API key" :)

File: zscaler/plugin.spec.yaml [hunk 9]
```diff
@@ -75,3 +75,26 @@ actions:
...
+  blacklist_url:
+    title: Blacklist URL
+    description: Blacklist or unblacklist a URL
+    input:
+      urls:
+        title: URLs
+        description: A given set of one or more URLs to update in the blacklist
+        type: "[]string"
+        required: true
+        example: ["example.com", "rapid7.com"]
```
Comment (jschipp-r7) [inline zscaler/plugin.spec.yaml:87]: These are domains, does this allow URLs? e.g. https://example.com

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]

File: zscaler/plugin.spec.yaml (hunk 9 above)
Comment (r7-kszczepanskagorna) [reply zscaler/plugin.spec.yaml:87]: In API they call it "url", but I test it and this endpoint accepts only domains. I will change description then for:
"A given set of one or more domain name to update in the blacklist"

Comment (jschipp-r7) [reply zscaler/plugin.spec.yaml:87]: @r7-kszczepanskagorna Gotcha, sorry, I just thought of a better way that would allow us to keep the same language as their API.

We can allow users to pass a URL like this: https://www.google.com or https://example.com/file.txt and we strip the uri prefix and the `/.*` suffix and then the action can handle both domains and URLs.

https://www.google.com -> google.com
 https://example.com/file.txt -> example.com

Then we can change the language back to URL in the descriptions

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]

**Review by jschipp-r7 — DISMISSED** [review]

File: zscaler/help.md [hunk 10]
```diff
@@ -44,11 +45,50 @@ Example input:
...
+
 ```
+
 ## Technical Details
 
 ### Actions
 
+#### Blacklist URL
+
+This action is used to add or remove  URLs from a blacklist.
```
Comment (jschipp-r7) [inline zscaler/help.md:57]:
```suggestion
This action is used to add or remove URLs from a blacklist. These URLs will appear in the "Blocked Malicious URLs" section on the Advanced Threats Policy page.
```

**Review by jschipp-r7 — APPROVED** [review]
